api/permissions_requests_handlers.go

```diff
@@ -0,0 +1,96 @@
+package api
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/topfreegames/Will.IAM/models"
+	"github.com/topfreegames/Will.IAM/usecases"
+	"github.com/topfreegames/extensions/middleware"
+)
+
+func permissionsRequestsCreateHandler(
+	prsUC usecases.PermissionsRequests,
+) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		l := middleware.GetLogger(r.Context())
+		pr := &models.PermissionRequest{}
+		if err := readBodyTo(r, pr); err != nil {
+			l.WithError(err).Error("failed to read body")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		saID, _ := getServiceAccountID(r.Context())
+		pr.ServiceAccountID = saID
+		if err := prsUC.WithContext(r.Context()).Create(pr); err != nil {
+			l.WithError(err).Error("failed to create permission request")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		if pr.ID == "" {
+			w.WriteHeader(http.StatusAccepted)
+			return
+		}
+		w.WriteHeader(http.StatusCreated)
+	}
+}
+
+func permissionsRequestsDenyHandler(
+	prsUC usecases.PermissionsRequests,
+) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		l := middleware.GetLogger(r.Context())
+		saID, _ := getServiceAccountID(r.Context())
+		prID := mux.Vars(r)["id"]
+		if err := prsUC.WithContext(r.Context()).Deny(saID, prID); err != nil {
+			l.WithError(err).Error("failed to deny permission request")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.WriteHeader(http.StatusAccepted)
+	}
+}
+
+func permissionsRequestsGrantHandler(
+	prsUC usecases.PermissionsRequests,
+) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		l := middleware.GetLogger(r.Context())
+		saID, _ := getServiceAccountID(r.Context())
+		prID := mux.Vars(r)["id"]
+		if err := prsUC.WithContext(r.Context()).Grant(saID, prID); err != nil {
+			l.WithError(err).Error("failed to grant permission request")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.WriteHeader(http.StatusAccepted)
+	}
+}
+
+func permissionsRequestsListOpenHandler(
+	prsUC usecases.PermissionsRequests,
+) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		l := middleware.GetLogger(r.Context())
+		saID, _ := getServiceAccountID(r.Context())
+		listOptions, err := buildListOptions(r)
+		if err != nil {
+			Write(
+				w, http.StatusUnprocessableEntity,
+				fmt.Sprintf(`{ "error": "%s"  }`, err.Error()),
+			)
+			return
+		}
+		prs, count, err := prsUC.WithContext(r.Context()).ListOpenRequestsVisibleTo(listOptions, saID)
+		if err != nil {
+			l.Error(err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		WriteJSON(w, 200, map[string]interface{}{
+			"count":   count,
```

Comment:
WDYT about having a struct with the fields here and marshaling it? Is it worth the trouble? Could it maybe be reused?

Ditto for line 81 (`{ "error": "..." }`)

Reply:
in what pkg would you place these structs? models? api?

Reply:
Depends. If they're meant to be reused, pick the place you think is most appropriate. If they're just local, defining close to where they're used is fine.

Overall I think you're better equip to make this call than I am.